lib/surface/api.ex

```diff
@@ -433,6 +495,36 @@ defmodule Surface.API do
     {:error, "invalid value for option :as in slot. Expected an atom, got: #{inspect(value)}"}
   end
 
+  defp validate_opt(:data, :upload, :accept, [_ | _]), do: :ok
+  defp validate_opt(:data, :upload, :accept, :any), do: :ok
+
+  defp validate_opt(:data, :upload, :accept, value) do
+    {:error,
+     "invalid value for option :accept. Expected a non empty list or :any, got: #{inspect(value)}"}
+  end
+
+  defp validate_opt(:data, :upload, :auto_upload, value) when not is_boolean(value) do
+    {:error, "invalid value for option :auto_upload. Expected a boolean, got: #{inspect(value)}"}
+  end
+
+  defp validate_opt(:data, :upload, :external, value) when not is_function(value, 2) do
+    {:error, "invalid value for option :external. Expected a function/2, got: #{inspect(value)}"}
+  end
+
+  defp validate_opt(:data, :upload, :progress, value) when not is_function(value, 3) do
+    {:error, "invalid value for option :progress. Expected a function/3, got: #{inspect(value)}"}
+  end
+
+  for option <- [:max_entries, :max_file_size, :chunk_size, :chunk_timeout] do
```

Comment:
Maybe we are missing some clarity in the code. We could (1) create 1 function for each option or (2) move the list into a guard clause. WDYT?

Reply:
In this case I would go with a guard clause, that should be simple and clear.

I've seen this pattern used in the Phoenix Html codebase like so:

```Elixir
for scheme <- @valid_uri_schemes do
  defp valid_string_destination!(unquote(scheme) <> _ = string, _context), do: string
end

defp valid_string_destination!(to, context) do
  if not match?("/" <> _, to) and String.contains?(to, ":") do
    raise ArgumentError, """
    unsupported scheme given to #{context}. In case you want to link to an
    unknown or unsafe scheme, such as javascript, use a tuple: {:javascript, rest}
    """
  else
    to
  end
end
```

But in this case multiple functions are needed to pattern match on the start of the string `unquote(scheme) <> _ = string`, which can't be done with a guard.